Refactor: Login Mechanism & resolve Cypress Test Build Issue

saisudan2003 opened the pull request:
## Description
Resolves: [Issue #209](https://github.com/jalantechnologies/boilerplate-mern/issues/209)
Resolves: [Issue #137](https://github.com/jalantechnologies/boilerplate-mern/issues/137)
Resolves: [Issue #225](https://github.com/jalantechnologies/boilerplate-mern/issues/225)

### Changes Implemented for Each Issue
**Issue #209 – Render the login page based on the provided props**
- authenticationMechanism introduced to control the current authentication method ('EMAIL' / 'PHONE')
  - authenticationMechanism: 'EMAIL' -> Email and password-based login
  - authenticationMechanism: 'PHONE' -> Phone number and OTP-based login

- Refactored routing logic in `src/apps/fontend/routes` folder:
  - `src/apps/frontend/routes/public.tsx`
    - conditionally mount the components based on the `authenticationMechanism` config variable.
      - authenticationMechanism: 'EMAIL' -> we mount `<Login />` and `<Signup />` -> Email based authentication
      - authenticationMechanism: 'PHONE' -> we mount `<PhoneLogin />` and `<OTPVerificationPage />` -> PhoneNumber based authentication
  - `src/apps/frontend/routes/auth-route.tsx` - deprecated


- refactored the the `PHONE_LOGIN` and `OTP` routes to improve readablity.
  - Removed the `PHONE_LOGIN`:
    - The `/login` route now dynamically renders either phone number-based or email-based login, based on the `authenticationMechanism` config variable.
  - Renamed `OTP` to `VERIFY_OTP` and refactored it as follows:
    - previous route: `/login?auth_mode=otp`
    - refactored route: `/otp/verify`

**Issue #137 – Add a link to Login/SignUp using OTP mode**
- Fixed the Button styling on the OTP authentication pages
- Adding a Link is now not required since we wish to have either a Email based authentication or a Phone Number based authentication and not both.

**Issue #225 – Fix (compose:e2e) Tests – Ensure Build Reflects 'NODE_CONFIG_ENV=e2e'**
- Modified `test:e2e` script in `package.json` to ensure Cypress tests run on the correct build:

    `"test:e2e": "cross-env NODE_ENV=production NODE_CONFIG_ENV=e2e npm run build && cross-env NODE_ENV=production NODE_CONFIG_ENV=e2e concurrently --success first --kill-others --kill-others-on-fail \"wait-on tcp:8080 && cypress run\" \"npm start\""`

- This ensures that the application is rebuilt with `NODE_CONFIG_ENV=e2e` before Cypress tests are executed.
- Prevents Cypress from testing against outdated builds.

## Database schema changes
NIL

## Tests
### Manual test cases run

-Verified login page renders correctly for both 'EMAIL' and 'PHONE' authenticationMechanism.

**Demo video: `authenticationMechanism: 'EMAIL'`**
This video depicts the login/signup flow when `authenticationMechanism: 'EMAIL'`
https://github.com/user-attachments/assets/9d81da01-5f97-43a6-9317-37b4a75e30f5


**Demo video: `authenticationMechanism: 'PHONE'`**
This video depicts the phoneNumber-OTP based login flow `authenticationMechanism: 'PHONE'` on localhost with the refactored `PHONE_LOGIN` and `OTP` routes.
https://github.com/user-attachments/assets/1e519626-e98f-4e2c-bd4b-dce4f3f13ead


-compose(e2e) checks pass when the `authenticationMechanism: 'PHONE'` in the default config which indicates that the cypress test are running on the build with `NODE_CONFIG_ENV=e2e` since the cypress tests expects username and password fields and for these tests to pass the application must be built with `NODE_CONFIG_ENV=e2e` which has `authenticationMechanism: 'EMAIL'`

YuvrajSinghJadon commented:
### Can you add what issue this PR is related to , in the PR descrption

YuvrajSinghJadon commented on src/apps/frontend/contexts/auth.provider.tsx:
```diff
@@ -80,7 +89,19 @@ const verifyOTPFn = async (
+  displayPhoneLoginOnWeb: false,
+  displayRegisterAccount: true,
+  currentLoginMethod: null as LoginMethod | null,
+};
```
- Consider setting an initial value for currentLoginMethod instead of null in initialLoginProps. Since the app already decides the login method based on screen width, initializing it to defaultMobileLogin or defaultWebLogin (depending on window.innerWidth) could prevent unnecessary state updates on first render. This would ensure the login page starts with the correct method without waiting for the useEffect update.

saisudan2003 replied:
Hey, this is a WIP PR, so these aren't the final changes. Still working through things. I'll link the PR on the Teams channel once it's ready for review.

YuvrajSinghJadon replied:
Yeah, I'm just adding suggestions that might help you.

govind-io commented on src/apps/backend/modules/account/account-service.ts:
```diff
@@ -42,6 +42,24 @@ export default class AccountService {
     return account;
   }
 
+  public static async getOrCreateAccountByPhoneNumberAndName(
```
Why do we need this function at first place? It should be two explicit functions `getAccountByPhoneNumber` and `createAccountByPhoneNumberAndName` . 

Right now the confusing part is:

1. Account already exists with a given phone number and the name is "Govind"
2. I try to create another account by passing in same number but a different name, it will just return an already created account with different name.

saisudan2003 replied:
I have attempted to build the (name+phoneNumber) based signup similar to the pre existing `getOrCreateAccountByPhoneNumber`.
The `getOrCreateAccountByPhoneNumber` (pre-existing) function either creates a new account with the provided phone number (if it doesn't exist in the DB) or returns an existing account with the provided phone number.

This was the reason to build the `getOrCreateAccountByPhoneNumberAndName` in a similar way.

saisudan2003 replied:
Got it, will rebuild it to throw an error if the phoneNumber already exists else create a new account and return it.

govind-io commented on src/apps/backend/modules/account/internal/account-writer.ts:
```diff
@@ -37,21 +37,24 @@ export default class AccountWriter {
-    const isValidPhoneNumber = phoneUtil.isValidNumber(
-      phoneUtil.parse(phoneNumber.toString()),
-    );
+    await this.validateAndCheckPhoneNumberExists(phoneNumber);
```
Let's try to avoid multiple responsibility functions, instead we should have two different functions here:
1. `validatePhoneNumber`
2. `CheckPhoneNumberExist` -> (ideally this should be just `getAccountByPhoneNumber`)

govind-io commented on src/apps/backend/modules/account/account-service.ts:
```diff
@@ -42,6 +42,24 @@ export default class AccountService {
+    phoneNumber: PhoneNumber,
+  ): Promise<Account> {
+    let account =
+      await AccountReader.getAccountByPhoneNumberOptional(phoneNumber);
```
This is confusing, what does it means to `getAccountByPhoneNumberOptional` ? What is optional here? Phone number? or getting an account? We should use clear and precise names for our functions.

saisudan2003 replied:
`getAccountByPhoneNumberOptional` is a pre existing function, not something I created in this PR.

govind-io commented on src/apps/backend/modules/account/internal/account-writer.ts:
```diff
@@ -95,4 +98,22 @@ export default class AccountWriter {
+    }
+    // check if account already exists with the given phone number
+    // this will throw an error if it does
+    await AccountReader.checkPhoneNumberNotExists(phoneNumber);
```
Readers job should be to read the data from db, it should not do any logical operations, those belongs to the service layer, reader should define something like: `getAccountByPhoneNumber` and if it returns null `account-service` should decide what to do with it.

saisudan2003 replied:
Sir `checkPhoneNumberNotExists` is a pre existing function, not something I created in this PR.

govind-io commented on src/apps/backend/modules/account/internal/account-writer.ts:
```diff
@@ -95,4 +98,22 @@ export default class AccountWriter {
     await AccountRepository.findByIdAndDelete(accountId);
   }
+
+  private static async validateAndCheckPhoneNumberExists(
```
Validation functions does not belongs to writer class, they should be part of the account-utils class

govind-io commented on src/apps/backend/modules/account/rest-api/account-controller.ts:
```diff
@@ -31,6 +31,12 @@ export class AccountController {
           username,
         );
+      } else if (firstName && lastName && phoneNumber) {
+        account = await AccountService.getOrCreateAccountByPhoneNumberAndName(
```
This should be `createAccountByPhoneNumberWithName`

govind-io commented on src/apps/backend/modules/account/rest-api/account-controller.ts:
```diff
@@ -31,6 +31,12 @@ export class AccountController {
           password,
           username,
         );
+      } else if (firstName && lastName && phoneNumber) {
```
Ideally `lastName` and `firstName` should be optional

jjalan commented on src/apps/backend/modules/account/account-service.ts:
```diff
@@ -30,8 +31,7 @@ export default class AccountService {
   ): Promise<Account> {
-    let account =
-      await AccountReader.getAccountByPhoneNumberOptional(phoneNumber);
+    let account = await AccountReader.getAccountByPhoneNumber(phoneNumber);
```
Why is there a change in formatting? We all should be using checked in formatting and linting rules.

jjalan commented on src/apps/backend/modules/account/account-service.ts:
```diff
@@ -42,6 +42,28 @@ export default class AccountService {
     return account;
   }
 
+  public static async createAccountByPhoneNumberAndName(
```
We already have `getOrCreateAccountByPhoneNumber` and there should not be a need for this.

jjalan commented on src/apps/backend/modules/account/internal/account-reader.ts:
```diff
@@ -54,18 +53,6 @@ export default class AccountReader {
 
   public static async getAccountByPhoneNumber(
     phoneNumber: PhoneNumber
```
I am not really expecting a change in backend, would like to understand why its needed

Killg0d commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -10,7 +10,8 @@ import {
 } from '../../../components';
 import routes from '../../../constants/routes';
-import { AsyncError } from '../../../types';
+import { getConfigValue } from '../../../helpers/config';
```
This directly imports getConfigValue, but the convention in helpers/config exposes everything under Config.
This breaks consistency and makes the code harder to refactor.

Usage is like import { Config } from '../../../helpers';
Config.getConfigValue\<string>('someKey');

Killg0d commented on src/apps/frontend/pages/authentication/phone-login/phone-login-form.tsx:
```diff
@@ -9,8 +10,10 @@ import {
-import { AsyncError } from '../../../types';
-import { ButtonKind, ButtonType } from '../../../types/button';
+import routes from '../../../constants/routes';
+import { getConfigValue } from '../../../helpers/config';
```
Same as above

Killg0d commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -23,6 +24,14 @@ interface LoginFormProps {
+  const currentLoginMethod = getConfigValue('currentLoginMethod');
+  const navigate = useNavigate();
+  useEffect(() => {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-enum-comparison
```
Is disabling lint a must here?

Killg0d reviewed (changes requested):
Left comments

jjalan commented on config/default.yml:
```diff
@@ -16,3 +16,6 @@ passwordResetToken:
   port: '8080'
+
+public:
+  currentLoginMethod: 'PHONE' #or 'EMAIL'
```
Can we please keep current login method as `EMAIL`.

Also, rename this property to be `authenticationMechanism`

jjalan commented:
When auth mechnaism is PHONE, there is a query string param in URL - auth_mode = otp, that should not be there.

jjalan commented on src/apps/frontend/pages/authentication/login/login-form.tsx:
```diff
@@ -23,6 +25,15 @@ interface LoginFormProps {
+  const navigate = useNavigate();
+  useEffect(() => {
+    if (currentLoginMethod === constant.PHONE_NUMBER_BASED_AUTHENTICATION) {
+      navigate(routes.SIGNUP);
```
It is not clear why ar we redirecting the user to signup page when user come to login page for phone base auth.

saisudan2003 commented on src/apps/frontend/pages/authentication/otp/otp-form.tsx:
```diff
@@ -88,9 +88,10 @@ const OTPForm: React.FC<OTPFormProps> = ({
-          type={ButtonType.SUBMIT}
           isLoading={isVerifyOTPLoading}
           kind={ButtonKind.PRIMARY}
+          size={ButtonSize.LARGE}
```
This change is associated with the Issue [#137](https://github.com/jalantechnologies/boilerplate-mern/issues/137) 

  - Fixed the Button styling on the OTP authentication pages
  - Adding a Link is now not required since we wish to have either a Email based authentication or a Phone Number based authentication and not both.

saisudan2003 commented:
The original routes for `PHONE_LOGIN` and `OTP` were as follows (src/apps/frontend/constants/routes.ts):
  - OTP: `'/login?auth_mode=otp'`
  - PHONE_LOGIN: `'/signup?auth_mode=otp'`
  
PHONE_LOGIN - refers to the route where we enter the phone number
OTP - refers to the route where we enter the OTP (although the exact route looks something like `${routes.OTP}&country_code=${encodedCountryCode}&phone_number=${formattedPhoneNumber}` )

Now since we removed the query params `auth_mode=otp` the refactored routes are:
  - OTP: `'/login'`
  - PHONE_LOGIN: `'/signup'`
  
Since we already have the following in the (src/apps/frontend/constants/routes.ts)
- `LOGIN: '/login'`
- `SIGNUP: '/signup'`
I removed the OTP and PHONE_LOGIN routes to avoid repetition.


This is the reason why we are redirecting the routes.SIGNUP for phone base auth.

jjalan reviewed (commented):
Routes should be:

```
Email:
/signup
/login

Phone:
/login
/otp/verify

```
